mypy/meet.py

```diff
@@ -285,15 +288,13 @@ def _type_object_overlap(left: ProperType, right: ProperType) -> bool:
     # Finally, we handle the case where left and right are instances.
 
     if isinstance(left, Instance) and isinstance(right, Instance):
-        if left.type.is_protocol and is_protocol_implementation(right, left):
-            return True
-        if right.type.is_protocol and is_protocol_implementation(left, right):
+        # First we need to handle promotions and structural compatibility for instances
+        # that came as fallbacks, so simply call is_subtype() to avoid code duplication.
+        if (is_subtype(left, right, ignore_promotions=ignore_promotions)
+                or is_subtype(right, left, ignore_promotions=ignore_promotions)):
```

Comment:
It's somewhat unfortunate we end up needing to perform a subtype check again here (we already call `is_proper_subtype` earlier in the method).

But as you said, it doesn't seem like there's a clean way of working around this without repeating some of the promotion logic anyways...